code/mob/living/critter/ai/sawflyai.dm

```diff
@@ -0,0 +1,212 @@
+				if(prob(20)) walk_rand(src,4)
+				owncritter.set_dir(get_dir(owncritter, holder.target)) //attack regardless
+				owncritter.hand_range_attack(holder.target, dummy_params)
+			else
+				//o no
+				frustration++
+			frustration++ //if frustration gets too high, the task is ended and re-evaluated
+		else
+			owncritter.set_dir(get_dir(owncritter, holder.target))
+			owncritter.hand_range_attack(holder.target, dummy_params)
+			SPAWN(10)
+				if(dist > target_range) // double check you're still in range
+					//replaced
+					if(!src.found_path)
+						src.found_path = get_path_to(holder.owner, holder.target, 18, 0)
+					if(src.found_path)
+						holder.move_to_with_path(holder.target, src.found_path, 1)
+						if(prob(20)) walk_rand(src,4)
+						owncritter.set_dir(get_dir(owncritter, holder.target))
+						owncritter.hand_range_attack(holder.target, dummy_params)
+					else
+						frustration++
+						//o no
+				else
+					owncritter.set_dir(get_dir(owncritter, holder.target))
+					owncritter.hand_range_attack(holder.target, dummy_params)
+
+
+
+
+/datum/aiTask/timed/targeted/sawfly_attack/get_targets()
+	. = list()
+	var/mob/living/critter/sawfly/owncritter = holder.owner
+
+	for (var/mob/living/C in view(owncritter,target_range))
+		if(C == owncritter) continue
+		if(istype(C, /mob/living/critter/sawfly)) continue
+		if (C.health < -50) continue
+		if (C.job == "Security Officer" || C.job == "Head of Security")
+			. = list(C) //found a secoff, just return that
+			return
+		if (C in owncritter.friends) continue
+		if (istraitor(C) || isnukeop(C) || isspythief(C)) // frens :)
+			boutput(C, "<span class='alert'> The sawfly's IFF system silently flags you as an ally! </span>")
+			owncritter.friends += C
+			continue
+		. += C //you passed all the checks it, now you get added to the list for consideration
+
+//chase behaviour - pick someone, run up to them, and stab em
+/datum/aiTask/sequence/goalbased/sawfly_chase_n_stab
+	name = "chasing"
+	weight = 15
+	max_dist = 12
+	can_be_adjacent_to_target = TRUE
+	var/found_path = null
+
+/datum/aiTask/sequence/goalbased/sawfly_chase_n_stab/New(parentHolder, transTask)
+	..(parentHolder, transTask)
+	add_task(holder.get_instance(/datum/aiTask/succeedable/sawfly_stab, list(holder)))
+
+/datum/aiTask/sequence/goalbased/sawfly_chase_n_stab/precondition()
+	return TRUE //put stuff here if you want them to not chase under some conditions
+
+/datum/aiTask/sequence/goalbased/sawfly_chase_n_stab/evaluate()
+	. = precondition() * weight * score_target(get_best_target(get_targets()))
+
+/datum/aiTask/sequence/goalbased/sawfly_chase_n_stab/on_tick()
+
+	if(!holder.target)
+		holder.target = get_best_target(get_targets())
+	..()
+
+/datum/aiTask/sequence/goalbased/sawfly_chase_n_stab/get_targets()
+	. = list()
+	var/mob/living/critter/sawfly/owncritter = holder.owner
+	for (var/mob/living/C in view(owncritter,max_dist))
+		if(C == owncritter) continue
```

Comment:
Ditto, also move the typecheck up. In general it's better to do checks in order of how computationally costly they are, and type checks are much cheaper than string checks